pkg/kubeadm/master/apiclient.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubemaster
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	//"k8s.io/kubernetes/pkg/client/unversioned" // TODO use a clientset
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+func CreateClientAndWaitForAPI(adminConfig *clientcmdapi.Config) (*clientset.Clientset, error) {
+	adminClientConfig, err := clientcmd.NewDefaultClientConfig(
+		*adminConfig,
+		&clientcmd.ConfigOverrides{},
+	).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("configuring the client...")
+
+	client, err := clientset.NewForConfig(adminClientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("polling API server forever...")
+
+	count := 0
+
+	wait.PollInfinite(500*time.Millisecond, func() (bool, error) {
+		count = count + 1
+		cs, err := client.ComponentStatuses().List(api.ListOptions{})
+		if err != nil {
+			return false, nil
+		}
+		if len(cs.Items) < 3 {
```

Comment:
The second check iterates over all components, is that good enough, or you
mean we should look the up by name? I think etcd names would be tricker in
an HA setting, so I left it as is for now...

On Wed, 31 Aug 2016, 22:18 Joe Beda, <notifications@github.com> wrote:

> In pkg/kubeadm/master/apiclient.go
> <https://github.com/kubernetes/kubernetes/pull/31221#discussion_r77075480>
> :
>
> > +	client, err := clientset.NewForConfig(adminClientConfig)
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	fmt.Println("polling API server forever...")
> > +
> > +	count := 0
> > +
> > +	wait.PollInfinite(500*time.Millisecond, func() (bool, error) {
> > +		count = count + 1
> > +		cs, err := client.ComponentStatuses().List(api.ListOptions{})
> > +		if err != nil {
> > +			return false, nil
> > +		}
> > +		if len(cs.Items) < 3 {
>
> This seems fragile. Not sure that we have anything better at this point.
> You could explicitly check for specific items.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31221/files/57096ca902c5c666d45df191037d9278972af1fd#r77075480>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS1cCYKou8A1uplN4RZ9kUdabKBT-ks5qle-8gaJpZM4JqsTH>
> .
>